sys/mips/timer.c

```diff
@@ -27,9 +29,22 @@ static bintime_t mips_timer_gettime(timer_t *tm);
 static uint64_t read_count(mips_timer_state_t *state) {
   SCOPED_INTR_DISABLED();
   state->count.lo = mips32_get_c0(C0_COUNT);
+
   /* detect hardware counter overflow */
-  if (state->count.lo < state->last_count_lo)
+  if (state->count.lo < state->last_count_lo) {
+    state->cntr_modulo +=
+      ((uint32_t)(-1) - state->last_count_lo) + state->count.lo;
     state->count.hi++;
+  } else {
+    state->cntr_modulo += state->count.lo - state->last_count_lo;
+  }
+
+  if (state->cntr_modulo >= state->timer.tm_frequency) {
+    state->cntr_modulo -= state->timer.tm_frequency;
+    state->sec++;
+  }
+  assert(state->cntr_modulo < state->timer.tm_frequency);
```

Comment:
When testing on QEMU 5.2.0, the kernel immediately panics on this assertion, because the counter value is too high at the beginning. We can fix this either by changing the `if` above to a `while`, or setting `C0_COUNT` to 0 in `mips_timer_start()`. I personally prefer the latter.

Reply:
Done.